keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -2049,6 +2046,247 @@ Unreserve is called in two scenarios:
 
 ### kubelet
 
+#### Communication between kubelet and resource kubelet plugin
+
+Resource kubelet plugins are discovered through the [kubelet plugin registration
+mechanism](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#device-plugin-registration). A
+new "ResourcePlugin" type will be used in the Type field of the
+[PluginInfo](https://pkg.go.dev/k8s.io/kubelet/pkg/apis/pluginregistration/v1#PluginInfo)
+response to distinguish the plugin from device and CSI plugins.
+
+Under the advertised Unix Domain socket the kubelet plugin provides the
+k8s.io/kubelet/pkg/apis/dra gRPC interface. It was inspired by
+[CSI](https://github.com/container-storage-interface/spec/blob/master/spec.md),
+with “volume” replaced by “resource” and volume specific parts removed.
+
+#### REST proxy
```

Comment:
/sig api-machinery auth
/cc @jpbetz @deads2k @enj 

For visibility to a `kube-apiserver <-- kubelet <-- client` proxied API surface proposal

The read part of this is similar in some ways to what was proposed in https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4188-kubelet-pod-readiness-api#proposed-api

1. the reads / watches are passed through to kube-apiserver with the kubelet's credentials
2. the API is a complete proxy, so client-go is usable (though tunneled over gRPC, which I'm not sure we've done before)
3. the API allows drivers to write through to kube-apiserver using the kubelet's credentials

A lot of the discussion around that proposal seems relevant to this one (sig-arch 2023-08-24, https://docs.google.com/document/d/1BlmHq5uPyBUDlppYqAAzslVbAO8hilgjqZUTaNXUhKM/edit#bookmark=id.vud1o04xj4iv, https://www.youtube.com/watch?v=toN7t_y4zCk&t=22m45s)

Reply:
> The read part of this is similar in some ways to what was proposed in https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4188-kubelet-pod-readiness-api#proposed-api

One key difference is that I am not proposing to add a new kubelet socket. I was worried about security implications around that. Instead, kubelet connects to plugins the same way as before and takes requests through one stream per plugin which is kept open by each plugin.

Reply:
Given a desire to read/write effectively from the kube-apiserver and the development of validatingadmissionpolicy + serviceaccount node claims + downward API for node name that can be used to self-restrict reads (for convenience) and enforce write requirements (must be from the same node), why is the proxy a better and more secure choice?

Reply:
I'm not familiar enough with that alternative to do a comparison. Can you point me to documentation?

The ask is to ensure that a DRA driver deployed as daemonset gets RBAC permissions which allow it to read/create/update ResourceSlice objects where the "nodeName" field is the same as the node on which the pod is running.

Reply:
For ResourceClaim, the ask is to ensure that the pod only gets read permission for ResourceClaims referenced by pods which have been scheduled to the node.

Reply:
@deads2k: I looked through https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy and https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/. It's not clear to me how I can connect the identity of the driver pod with the node that it is running on and that identity with rules that restricts what that pod can access.

The node authorizer uses hand-crafted Go code for this.

Reply:
See https://github.com/kubernetes/kubernetes/pull/124711 for the VAP+SA example David is referring to.

Note that admission only covers writes, so reads cannot be restricted in this way.

Reply:
This seems to be the key part:
`object.metadata.name == request.userInfo.extra[\"authentication.kubernetes.io/node-name\"][0])`

Does `object` provide access to any field?

Is it possible to extend what gets added to the credentials? If there was a `dra.kubernetes.io/driver-name`, then the same check could be done for the `driverName` field in ResourceSlice. That would be better than what the REST proxy can do.

I'm fine with dropping the read check on ResourceClaim. Node authorizer also has a "loop hole" there because it does limit what data gets returned by watches.